Annotator: replace std::pair with AnyCellmlElement

agarny opened the pull request:
Closes #748.

kerimoyle committed:
Created AnyItemNew class

kerimoyle committed:
Change AnyItem std::pair into AnyItem class in Annotator, still need to do Issues

kerimoyle committed:
Merge branch 'develop' into 748-any-item-type

kerimoyle committed:
python coverage

kerimoyle committed:
Includes in types.cpp

kerimoyle committed:
Python tests for AnyItem (WIP)

kerimoyle committed:
Merge branch 'develop' into 748-any-item-type

kerimoyle committed:
Merge branch 'develop' into 748-any-item-type

agarny committed:
Merge branch 'main' into pr/783.

agarny committed:
Merge branch 'main' into issue748.

agarny committed:
Merge branch 'main' into issue748.

agarny committed:
Merge branch 'main' into issue748.

agarny committed:
Merge branch 'main' into issue748.

hsorby commented on src/api/libcellml/types.h:
```diff
@@ -235,6 +236,86 @@ class LIBCELLML_EXPORT VariablePair
+    AnyItem(AnyItem &&rhs) noexcept = delete; /**< Move constructor. */
+    AnyItem &operator=(AnyItem rhs) = delete; /**< Assignment operator. */
+
+    static AnyItemPtr create() noexcept;
```
Requires Doxygen documentation.

agarny replied:
Not relevant anymore.

hsorby commented on src/api/libcellml/types.h:
```diff
@@ -235,6 +236,86 @@ class LIBCELLML_EXPORT VariablePair
+    AnyItem &operator=(AnyItem rhs) = delete; /**< Assignment operator. */
+
+    static AnyItemPtr create() noexcept;
+    /**
```
New line here.

agarny replied:
Done in commit 98458b98.

hsorby commented on src/api/libcellml/types.h:
```diff
@@ -235,6 +236,86 @@ class LIBCELLML_EXPORT VariablePair
+     * @return A smart pointer to a @ref AnyItem object.
+     */
+    static AnyItemPtr create(CellmlElementType type, const std::any &item) noexcept;
+    // KRM
```
Remove

agarny replied:
Done in commit 6e7e6831.

hsorby commented on src/api/libcellml/types.h:
```diff
@@ -235,6 +236,86 @@ class LIBCELLML_EXPORT VariablePair
+    static AnyItemPtr create(CellmlElementType type, const UnitPtr &item) noexcept;
+    static AnyItemPtr create(CellmlElementType type, const UnitsPtr &item) noexcept;
+    static AnyItemPtr create(CellmlElementType type, const VariablePtr &item) noexcept;
+    static AnyItemPtr create(CellmlElementType type, const VariablePairPtr &item) noexcept;
```
All of these require Doxygen documentation.

agarny replied:
Not relevant anymore.

hsorby commented on src/bindings/interface/types.i:
```diff
@@ -198,7 +224,132 @@ Provides support for shared pointers declared in types.h.
+%ignore libcellml::AnyItem::create;
 
-%include "libcellml/types.h"
+%include "libcellml/types.h"
```
New line

agarny replied:
Done in commit e72e60cb.

hsorby commented on src/types.cpp:
```diff
@@ -113,4 +118,157 @@ bool VariablePair::isValid() const
+struct AnyItem::AnyItemImpl
+{
+    std::any mItem = nullptr; /**< std::any item cast for the item.*/
+    CellmlElementType mType = CellmlElementType::UNDEFINED; /**< Type for the item.*/
```
space in `.*/` should be `. */`

agarny replied:
Done in commit ffec2304.

hsorby commented on tests/any_item/any_item.cpp:
```diff
@@ -0,0 +1,35 @@
+    auto item = libcellml::AnyItem::create(libcellml::CellmlElementType::MODEL, std::any(model));
+    EXPECT_EQ(libcellml::CellmlElementType::MODEL, item->type());
+    EXPECT_EQ(model, std::any_cast<libcellml::ModelPtr>(item->item()));
+}
```
Needs test for all API of the class `setItem` and `setType`.

agarny replied:
I have removed `setItem()` and `setType()` from the `AnyCellmlElement` API. Indeed, I reckon that a user should only ever consume an `AnyCellmlElement` object, not modify its contents. The latter should only be made possible internally (by the `Annotator` class).

agarny replied:
Not relevant anymore.

hsorby reviewed (changes requested):
A few things need to be completed.

I would also like to ask should the class be called `AnyCellmlElement` and not `AnyItem`.

agarny committed:
Merge branch 'main' into issue748

agarny committed:
Removed trailing spaces here and there.

agarny committed:
Types: added a new line to make the code more readable.

agarny committed:
Types: removed unneeded comment.

agarny committed:
Types: added trailing empty line.

agarny committed:
Types: cleaned up a couple of comments.

agarny committed:
Renamed AnyItem to AnyCellmlElement.

agarny commented:
> A few things need to be completed.

Yes, so far this PR is really just about getting it back on track.

> I would also like to ask should the class be called `AnyCellmlElement` and not `AnyItem`.

Done in commit 63f411db.

agarny committed:
Renamed AnyItem to AnyCellmlElement.

Qt Creator hadn't done a very good job of it!

agarny committed:
Renamed AnyItem to AnyCellmlElement.

agarny committed:
Some minor cleaning up.

agarny committed:
Python: some work on AnyCellmlElement.

agarny committed:
Python: addressed some coverage issues.

agarny committed:
Some minor cleaning up.

agarny committed:
AnyCellmlElement: the `setItem()` and `setType()` shouldn't be part of the API.

A user should only ever consume an `AnyCellmlElement` object, not change its item and/or type.

agarny committed:
Annotator: updated our Python tests.

agarny committed:
Some minor cleaning up.

agarny committed:
Annotator interface: removed the case where an argument is a list or a tuple.

Does such a case ever occur?

agarny committed:
Don't expose std::any in the API.

agarny commented:
Commit 6ed25db contains an awful lot of changes. Something that I really didn't like was that we were returning `std::any` from our API. As we know, this causes issues with SWIG, resulting in various hacks to get around that problem. So, now, although we keep on using `std::any` internally (it's indeed quite convenient), we don't return `std::any` from our API anymore. This should therefore make things easier on the SWIG (yet to be done).

Also, while doing this, I have removed the ability to create an `AnyCellmlElement` object from the API. I just don't understand why we allowed this in the first place. Such an object should only ever be created internally. As it happens, it's the same story with `Issue`. Why would someone using the API want to create an `Issue` object? So, yes, we can't now create those objects from the API. Unsurprisingly, this makes things much easier, both on the C++ front and on the SWIG side of things (at least, it should since I have yet to update that side of the library).

agarny committed:
Don't expose std::any in the API.

agarny commented:
@hsorby and @nickerso: this now ready for review.

agarny committed:
Merge branch 'main' into issue748.

agarny committed:
Merge branch 'main' into issue748.

hsorby committed:
Strip encapsulation and mapVariables etc. from issue.  Allow assignId the generic method.  Combine get methods into one.

hsorby committed:
Remove link component entity units from model class.

hsorby committed:
Chnage CellMLElement Type to use variablePair and not connection or map variables.

hsorby committed:
Revert to using assignId.

hsorby committed:
Remove assignId for AnyCellMLElementType.

agarny committed:
Merge branch 'main' into issue748.

hsorby committed:
Use all the assignId overloaded methods.

hsorby committed:
Merge branch 'issue748' of github.com:agarny/libcellml into agarny/issue748

agarny committed:
Merge branch 'main' into issue748.

hsorby committed:
Merge in changes from 'agarny/issue748' into 'agarny/issue748' fixing conflicts.

agarny committed:
CellmlElementType: renamed `UNIT` to `UNITS_ITEM`.

agarny committed:
Annotator: renamed `exists()` to `idCount()`.

agarny committed:
Some minor cleaning up.

agarny committed:
Reverted "CellmlElementType: renamed `UNIT` to `UNITS_ITEM`."

Indeed, CellmlElementType is a direct relationship to the XML elements defined in the specification.

agarny committed:
Code style: put "||" at the beginning of a line.

agarny committed:
Removed unneeded code.

agarny committed:
Make ClangFormat happy.

agarny committed:
Some minor cleaning up.

agarny committed:
AnyCellmlElement: removed unneeded pimpl methods.

agarny committed:
Some minor cleaning up.

agarny committed:
Removed commented-out code.

agarny committed:
Annotator: slight updates to our API documentation.

agarny committed:
Reverted to using assignId().

agarny committed:
Some minor cleaning up.

agarny committed:
CellmlElementType: order our use of it in if/case statements.

agarny committed:
Some minor cleaning up.

agarny committed:
Interfaces: removed unneeded methods from the Issue interface.

agarny committed:
Interface: reordered a method for AnyCellmlElement.

agarny committed:
Interfaces: removed unneeded methods from the Issue class for our JavaScript bindings.

agarny committed:
Python: code formatting.

agarny committed:
Replaced "[c|C]omponentRef" with "[c|C]omponentEncapsulation" when not dealing directly with component_ref.

agarny committed:
Some minor cleaning up.

agarny committed:
Annotator: various minor fixes to our API documentation.

agarny committed:
Annotator interface: some minor reordering.

Makes it easier to compare with the main branch.

agarny committed:
Utilities: some minor reordering.

Makes it easier to review.

agarny committed:
Annotator: updated our Python tests.

Since we are keeping assignId().

agarny committed:
Annotator: updated our Python tests.

agarny committed:
Annotator: updated our C++ tests.

Since we are keeping assignId().

agarny commented:
@hsorby & @nickerso: I believe this is once again ready for review.

agarny committed:
Make MSVC and Clang happy.

... and ClangFormat too while we are at it!

agarny committed:
Make coverage happy.

agarny committed:
Merge branch 'main' into issue748.

agarny committed:
Merge branch 'main' into issue748.

hsorby committed:
Merge in changes from 'prime/main' fixing conflicts.

hsorby committed:
Merge in changes from 'agarny/issue748' fixing conflicts.

hsorby committed:
Improve documentation in AnyCellMLElement.

hsorby committed:
Remove createCyclicDependencyIssue from private Issue class.

hsorby committed:
Remove resetValue and testValue from private AnyCellmlElement API.

agarny committed:
Reverted the last 5 commits.

i.e. commits 5f039a39, fc9e7acb, ff676997, a59037d1, and 772ab349.

agarny committed:
AnyCellMLElement: fixed the documentation.

agarny committed:
AnyCellmlElement: allow the creation of an object through its private implementation.

agarny committed:
AnalyserModel: allow the creation of an object through its private implementation.

agarny committed:
AnalyserVariable: allow the creation of an object through its private implementation.

agarny committed:
AnalyserInternalVariable: allow the creation of an object.

agarny committed:
AnalyserInternalEquation: allow the creation of an object.

agarny committed:
Issue: removed createCyclicDependencyIssue() from the private implementation.

agarny commented:
Ok, back on track. I have also incorporated some of @hsorby's changes. However:
- I can't see why, in `Validator`, we would use `setVariablePair()` rather than `setMapVariables()` and `setConnection()`. This is internal code and makes it easier for us to read/maintain/etc.
- Why get rid of `resetValue()` and `testValue()` in the `Issue` class? Not surprisingly (?), this breaks our annotator tests.

Anyway, this is now ready for (re-)review.

hsorby commented:
The main reason for getting rid of `setMapVariables` is so that we don't get it in the API.  I know it is internal but thought it might creep into the API if we kept it.

The issue class should for my thinking return the thing that it is we don't have a `testValue` object so we shouldn't be asking for one and getting a `Reset`.

hsorby commented on src/utilities.h:
```diff
@@ -656,11 +656,11 @@ bool traverseComponentEntityTreeLinkingUnits(const ComponentEntityPtr &component
  *  Utility function used when linking units names to their corresponding @c Units items.
```
@ref Units, not @c.

agarny replied:
I am inclined to leave it for another issue? I mean, I just searched for `@c Units` in our codebase and there are many instances of it (and other similar ones).

agarny replied:
Happy to change this particular one though, if you really insist. :)

hsorby replied:
Agreed that we should try and fix this in bulk. Maybe a session during a meeting? perhaps.  I have been fixing them up if I notice them for any functions or methods that I change.  I won't insist but I think it would be better to have it done than not.

agarny replied:
Done in commit b904f477.

hsorby commented on src/utilities.h:
```diff
@@ -630,16 +630,16 @@ std::string makeUniqueId(IdList &idList);
  *
  * @param component The component to check.
- * @param issueList A vector of @c IssuePtr items in which unlinked units are recorded for reporting.
+ * @param descriptionList A @c std::vector of @c std::string items in which unlinked units are recorded for reporting.
```
This is not what a description list is.

agarny replied:
The description was "wrong" in the first place, as are some other descriptions (just search for `descriptionList` and you will find one for `traverseComponentEntityTreeLinkingUnits()`). Anyway, happy to change the documentation for this (and the other) `descriptionList` if you tell me what you would like to see.

hsorby replied:
A @c std::vector of variables paired with a description.

agarny replied:
Done in commit 4090dfc6.

hsorby commented on tests/validator/validator.cpp:
```diff
@@ -51,10 +51,13 @@ TEST(Validator, unnamedModel)
+    const std::vector<std::string> expectedUrls = {
+        "https://cellml-specification.readthedocs.io/en/latest/reference/formal_and_informative/specB01.html?issue=MODEL_NAME",
+    };
     libcellml::ValidatorPtr validator = libcellml::Validator::create();
```
To be consistent we should have an empty line here.

agarny replied:
Done in commit a06be2cd.

hsorby commented on src/utilities.h:
```diff
@@ -630,16 +630,16 @@ std::string makeUniqueId(IdList &idList);
  * Function to support linking units names to their corresponding @c Units items.
```
@ref Units, not @c.

agarny replied:
See https://github.com/cellml/libcellml/pull/836#discussion_r634778487.

agarny replied:
Done in commit b904f47.

hsorby commented on src/utilities.cpp:
```diff
@@ -1292,19 +1282,19 @@ bool linkComponentVariableUnits(const ComponentPtr &component, std::vector<Issue
 bool traverseComponentEntityTreeLinkingUnits(const ComponentEntityPtr &componentEntity)
 {
-    std::vector<IssuePtr> issueList;
+    DescriptionList issueList;
```
`issueList` is a bad variable name.

agarny replied:
Nothing to do with this PR as such, but ok done in commit b41a9e7.

hsorby commented on src/types.cpp:
```diff
@@ -73,14 +68,20 @@ size_t UnitsItem::index() const
-    }
-
-    return false;
+    return mPimpl->mIndex < mPimpl->mUnits.lock()->unitCount();
```
I think I could cause this to segfault.

agarny replied:
Not sure who changed the original code to this, but if I did then mea culpa. I agree that this is not good. I have therefore reverted to the original code (see commit 3d375612). However, I am now wondering whether we are still going to get 100% coverage... could it be the reason that someone/I switched to the code above?

agarny replied:
I can indeed confirm that we don't have 100% coverage anymore. Ok, need to find a way to "segfault" this now...

agarny replied:
Looking into it, `UnitsItem` is very much an internal type and, yet, it's in our API...

agarny replied:
Ok, maybe not since it's (only) needed by the `Annotator` class.

agarny replied:
Note that coverage has been addressed in commit 9b2745ae. However, this highlights another issue which is that the API allows us to create an `UnitsItem` while it clearly shouldn't. So... we should do what we have for the `Issue` class, i.e. only allow the creation of of `Issue` instances from our internal codebase. Am going to do that...

agarny replied:
First, commit 70796228 implements a better way of testing coverage for `UnitsItem::isValid()`.

agarny replied:
I guess that `UnitsItem` is similar to `VariablePair`, i.e. a convenience class. So, ok, I am leaving things as-is then, which means that I have reverted commit 7079622.

hsorby commented on src/model_p.h:
```diff
@@ -0,0 +1,40 @@
+    std::vector<UnitsPtr>::const_iterator findUnits(const std::string &name) const;
+    std::vector<UnitsPtr>::const_iterator findUnits(const UnitsPtr &units) const;
+
+    bool equalUnits(const ModelPtr &other) const;
```
Should have Doxygen documentation.

agarny replied:
Do we seriously need to add documentation every of our single pimpl methods? As the ones writing the code, I would hope that our method and parameter names are clear enough to understand without having to document them. That's also what our review process is about: ensure that our code is clear.

Anyway, done in commit 9084b839.

hsorby replied:
In my opinion yes, we should have documentation for every internal method.

hsorby commented on src/importer.cpp:
```diff
@@ -193,21 +195,21 @@ bool Importer::ImporterImpl::checkComponentForCycles(const ComponentPtr &compone
-            issue->setReferenceRule(Issue::ReferenceRule::IMPORTER_MISSING_COMPONENT);
+            auto issue = Issue::IssueImpl::create();
+            issue->mPimpl->setDescription("Component '" + component->name() + "' imports a component named '" + component->importReference() + "' from the model imported from '" + resolvingUrl + "'. The component could not be found.");
+            issue->mPimpl->setLevel(Issue::Level::ERROR);
```
Redundant.

agarny replied:
Agreed, fixed in commit c2923e4f.

hsorby commented on src/importer.cpp:
```diff
@@ -193,21 +195,21 @@ bool Importer::ImporterImpl::checkComponentForCycles(const ComponentPtr &compone
-            issue->setReferenceRule(Issue::ReferenceRule::IMPORTER_NULL_MODEL);
+            auto issue = Issue::IssueImpl::create();
+            issue->mPimpl->setDescription("Component '" + component->name() + "' requires a model imported from '" + resolvingUrl + "' which is not available in the importer.");
+            issue->mPimpl->setLevel(Issue::Level::ERROR);
```
Redundant.

agarny replied:
Should have chcked the whole thing first. Ok, fixed this one and other similar ones in commit 7a25f5ed.

hsorby commented on src/importer.cpp:
```diff
@@ -155,21 +157,21 @@ bool Importer::ImporterImpl::checkUnitsForCycles(const UnitsPtr &units, History
-        issue->setReferenceRule(Issue::ReferenceRule::IMPORTER_MISSING_UNITS);
+        auto issue = Issue::IssueImpl::create();
+        issue->mPimpl->setDescription("Units '" + units->name() + "' imports units named '" + units->importReference() + "' from the model imported from '" + resolvingUrl + "'. The units could not be found.");
+        issue->mPimpl->setLevel(Issue::Level::ERROR);
```
Redundant.

agarny replied:
See commit 7a25f5ed.

hsorby commented on src/importer.cpp:
```diff
@@ -155,21 +157,21 @@ bool Importer::ImporterImpl::checkUnitsForCycles(const UnitsPtr &units, History
-        issue->setReferenceRule(Issue::ReferenceRule::IMPORTER_NULL_MODEL);
+        auto issue = Issue::IssueImpl::create();
+        issue->mPimpl->setDescription("Units '" + units->name() + "' requires a model imported from '" + resolvingUrl + "' which is not available in the importer.");
+        issue->mPimpl->setLevel(Issue::Level::ERROR);
```
Redundant.

agarny replied:
See commit 7a25f5ed.

hsorby commented on src/api/libcellml/model.h:
```diff
@@ -39,6 +39,8 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
 #endif
 {
+    friend class Parser;
+
```
Not required.

agarny replied:
Agreed, fixed in commit bd383374.

hsorby reviewed (changes requested):
First review pass.

hsorby committed:
Remove testValue() and resetValue() from AnyCellmlElement.

agarny committed:
Merge branch 'main' into issue748.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Types: prevent a potential segfault.

agarny committed:
Model: added some documentation to a Pimpl method.

agarny committed:
Importer: removed some redundant code.

agarny committed:
Removed some redundant code when creating an error-type issue.

agarny committed:
Model: no need for Parser to be a friend class.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Make our coverage tests happy.

agarny committed:
Utilities: replaced "@c Units" with "@ref Units".

agarny committed:
Units test: proper way of testing coverage for UnitsItem::isValid().

agarny committed:
Types: slight improvement to our API documentation.

agarny committed:
Revert "Units test: proper way of testing coverage for UnitsItem::isValid()."

This reverts commit 70796228ad21f44d7962e03b0f2b78a455c323d4.

hsorby commented on tests/units/units.cpp:
```diff
@@ -2731,3 +2731,28 @@ TEST(Units, circularImportDeeperLevelBaseUnits)
+    }
+
+    EXPECT_FALSE(unitsItem->isValid());
+}
```
If this is purely for coverage reasons then it belongs in the coverage test.

agarny replied:
FWIW, this test has been superseded by commit 157189f5. However, it has now been moved to the coverage test file (see commit 172ae199).

agarny committed:
Utilities: slight improvement to our documentation.

agarny commented:
Ok, I think I have addressed all of @hsorby's comments. So, ready for another round of reviews.

agarny committed:
Moved our coverage test for UnitsItem to the correct place.

hsorby commented on src/api/libcellml/annotator.h:
```diff
@@ -114,13 +114,13 @@ class LIBCELLML_EXPORT Annotator: public Logger
      * @param index The index of the item to return from the list of items with @p id.
      *
-     * @return An @c AnyItem item as described in item(const std::string &).
+     * @return An @c AnyCellmlElement item as described in item(const std::string &).
```
`@ref`

hsorby replied:
Done already.

hsorby commented on src/api/libcellml/annotator.h:
```diff
@@ -114,13 +114,13 @@ class LIBCELLML_EXPORT Annotator: public Logger
     /**
      * @brief From a list of items in the stored model with the given @p id string,
-     *        this method returns a @c ComponentPtr in the @p index position, if it exists.
+     *        this method returns a @c Component in the @p index position, if it exists.
```
`@ref`

hsorby replied:
Done already.

hsorby commented on src/api/libcellml/annotator.h:
```diff
@@ -133,32 +133,13 @@ class LIBCELLML_EXPORT Annotator: public Logger
      * @param index The position of an item within the list of items with the given @p id to retrieve.
      *
-     * @return A @c ComponentPtr on success otherwise @c nullptr.
+     * @return A @c Component on success otherwise @c nullptr.
```
`@ref`

hsorby replied:
Done already.

hsorby committed:
Replace use of `@c` with `@ref` for libCellML types.

hsorby commented on src/api/libcellml/annotator.h:
```diff
@@ -89,23 +89,23 @@ class LIBCELLML_EXPORT Annotator: public Logger
      *
-     * @return An @c AnyItem item (as described above).
+     * @return An @ref AnyCellmlElement item (as described above).
      */
```
I don't think this documentation is accurate any more.

agarny replied:
Done in commit c28817c1.

hsorby reviewed (changes requested):
I think the documentation needs a little look.

I have one more thing to check now.

agarny committed:
Annotator: updated the API documentation for the item() method.

hsorby commented on src/api/libcellml/annotator.h:
```diff
@@ -66,46 +66,37 @@ class LIBCELLML_EXPORT Annotator: public Logger
+     *    @ref VariablePair.
+     *
+     * @overload
+     *
```
This method is not the overloaded one.

agarny replied:
Argh, of course. Fixed in commit 0c635993.

agarny committed:
Some minor cleaning up.

nickerso commented on src/annotator.cpp:
```diff
@@ -437,57 +457,71 @@ void Annotator::setModel(const ModelPtr &model)
 }
 
-AnyItem Annotator::item(const std::string &id)
+bool Annotator::AnnotatorImpl::canReturnItem(const std::string &id, size_t index) const
```
its hard to tell with documentation to know what this method should be doing, but the behaviour of `canReturnItem` is different to the previous `exists` method. Not at all clear why an index is even needed if this is now going to error out with non-unique ids.

nickerso replied:
oops, it doesn't error out but return's `true` yet still sets an issue? This behaviour seems weird...

hsorby replied:
I think this is down to the philosophy of the `Annotator`.  The idea is that `id`s should be unique and if they aren't then the user should be told about it.

hsorby replied:
The name `canReturnItem` is more of a hangover from when there was a `std::pair` involved.  I think now it makes more sense to call this method `exists` again.

nickerso replied:
> I think this is down to the philosophy of the Annotator. The idea is that ids should be unique and if they aren't then the user should be told about it.

but this internal API change is not reflected in the current public API documentation or philosophy...or have I missed something?

hsorby replied:
I have made some changes to clarify this.

nickerso reviewed (changes requested):
Seems pretty good, but I think I'm missing something. Can you please explain, perhaps by adding some documentation to the `idCount` and `canReturnItem` methods in the annotator implementation, describing why that approach is better than the previous simple `exists` method. I'm not seeing how that is a good change...

hsorby commented on src/annotator.cpp:
```diff
@@ -437,57 +457,71 @@ void Annotator::setModel(const ModelPtr &model)
-    if ((mIdList.count(id) <= index) || (mIdList.count(id) == 0)) {
+
+    auto count = mIdList.count(id);
+    if ((count <= index) || (count == 0)) {
```
I believe the `(count == 0)` part of this condition is redundant.

hsorby committed:
Clarify and simplify code around testing for id existence in the Annotator.

hsorby committed:
Fix coverage issue with Annotator::AnnotatorImpl::exists().

nickerso commented on src/api/libcellml/annotator.h:
```diff
@@ -70,42 +70,36 @@ class LIBCELLML_EXPORT Annotator: public Logger
+     * the given @p id string and @p index.  The default @p index is zero if none is given.
      *
-     * See item(const std::string &) for a full breakdown of the @ref AnyCellmlElement return value.
+     * Errors are logged if:
```
issues not errors?

nickerso reviewed (changes requested):
Thanks @hsorby - these changes definitely make it clearer what that method is doing. But I still can't see how the introduction of this issue for non-unique IDs is an improvement? There shouldn't be an issue created if I ask for any item with an index greater than 0 as I clearly know that there are duplicate IDs. Adding an issue every time I do that just clogs up the annotator with issues that I'd need to check and ignore...or more likely just never check for issues as I know that I'm going to ignore them all every time.

hsorby committed:
Modify default so we can check when an index is not set and add an issue if required.

agarny commented on src/annotator.cpp:
```diff
@@ -512,22 +517,29 @@ bool Annotator::AnnotatorImpl::exists(const std::string &id, size_t index) const
         return false;
     }
 
+
```
Why this empty line?

hsorby replied:
5d65515

agarny commented on src/annotator.cpp:
```diff
@@ -512,22 +517,29 @@ bool Annotator::AnnotatorImpl::exists(const std::string &id, size_t index) const
 {
-    return mPimpl->exists(id, index) ? std::move(items(id)[index]) : AnyCellmlElement::AnyCellmlElementImpl::create();
+    auto privateIndex = realIndex(index);
+    return mPimpl->exists(id, index) ? std::move(items(id)[privateIndex]) : AnyCellmlElement::AnyCellmlElementImpl::create();
```
Why not simply:
```c++
return mPimpl->exists(id, index) ?
           std::move(items(id)[realIndex(index)]) :
           AnyCellmlElement::AnyCellmlElementImpl::create();
```

hsorby replied:
5d65515

agarny commented on src/annotator.cpp:
```diff
@@ -574,7 +586,8 @@ ComponentPtr Annotator::component(const std::string &id, size_t index)
-        auto i = items(id).at(index);
+        auto privateIndex = realIndex(index);
+        auto i = items(id).at(privateIndex);
         return i->component();
```
Why not simply (same elsewhere):
```c++
return items(id).at(realIndex(index))->component();
```
I know you do, but I really don't see the benefit of all those temporary variables.

hsorby replied:
5d65515

hsorby committed:
Remove use of temporary variable from Annotator methods.

hsorby committed:
Differentiate between asking for an item with an id and asking for a specific item with id.

hsorby committed:
add mPimpl->update(); to read methods in Annotator class.

hsorby committed:
Fix documentation on Annotator get * methods.

hsorby committed:
Rework the way issues are generated and when for the annotator.

hsorby committed:
Simplify the retrieval of objects from the annotator.

hsorby committed:
Fix indentation in annotator tests for clang format.

hsorby committed:
Correct the documentation for some annotator methods.

nickerso commented:
@agarny - will let you hit merge if you are still happy with this one.
